internal/pkg/repository/repository.go

```diff
@@ -0,0 +1,70 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// Package repository provides support for building and pushing images to a repository.
+package repository
+
+import (
+    "fmt"
+)
+
+// ContainerLoginBuildPusher provides support for logging in to repositories, building images and pushing images to repositories.
+type ContainerLoginBuildPusher interface {
+    Build(uri, path, imageTag string, additionalTags ...string) error
+    Login(uri, username, password string) error
+    Push(uri, imageTag string, additionalTags ...string) error
+}
+
+// Registry gets information of repositories.
+type Registry interface {
+    RepositoryURI(name string) (string, error)
+    Auth() (string, string, error)
+}
+
+// Repository builds and pushes images to a repository.
+type Repository struct {
+    repositoryName string
+    registry Registry
+
+    uri string
+}
+
+// New instantiates a new Repository.
+func New(name string, registry Registry) (*Repository, error){
+    uri, err := registry.RepositoryURI(name)
+    if err != nil {
+        return nil, fmt.Errorf("get repository URI: %w", err)
+    }
+
+    return &Repository{
+        repositoryName: name,
+        uri:            uri,
+        registry:       registry,
+    }, nil
+}
+
+// BuildAndPush builds the image from Dockerfile and pushes it to the repository with tags.
+func (r *Repository) BuildAndPush(docker ContainerLoginBuildPusher, dockerfilePath string, tag string, additionalTags ...string) error {
+    if err := docker.Build(r.uri, dockerfilePath, tag, additionalTags...); err != nil {
+        return fmt.Errorf("build Dockerfile at %s: %w", dockerfilePath, err)
+    }
+
+    username, password, err := r.registry.Auth()
```

Comment:
hmm any reason why we don't wrap username and password in a struct?

Reply:
I think they could be wrapped but the struct shouldn't be declared in an ecr package, because username/password aren't ecr-specific authentication fields. This then posts a question of where to put the struct💭  

This is initiated from this conversation: [remove dependency from ecr pkg](https://github.com/aws/copilot-cli/pull/1161#discussion_r456840025).

Reply:
ah nvm then. This should be the limitation for golang.